### DISCUSSION TITLE
Schematron

### DISCUSSION THREAD

**Discussion by Devasta** [ballsteve/xrust#76] [General]
Hi Steve!

I'm looking at getting Schematron support built into the project, using the transforms provided by the SchXslt project. I figure it'd be a nice feature to provide while also a demo that xrust can do complex transforms.

There are a couple of things we'll need to do before we can consider merging:

1. There are two transforms to run, one to compile the schema to a transform, the other to run that transform against your document.
2. On top of that, there are two versions of SchXslt, depending on whether or not the schema has an attribute "queryBinding" set for xslt2/3 or not. We'll need to check that before determining which one to run, or we could expose both to the user and have them decide?
3. Need to take the generates output and decide how we'll return the list of errors if there is a validation failure.
4. Currently I have coded the external file functions for xsl:include and xsl:import as being local to the filesystem, but need to consider how it would be used in other situations like in WASM.
5. Need to update our documentation to specifically mention that we are using SchXslt. Have included their license file in with the transforms.

xrust::validators::schematron::validate_schematron() is the function.
 
Hitting a bit of a wall at the moment, every time I attempt to run I get the error "already borrowed: BorrowMutError
thread 'conformance::schematron::svrl1::schematron_svrl_01' panicked at src\trees\smite.rs:475:16:"

Has me stumped at the moment, but once I figure out what the issue is on that, I can get it more built out. (dev_schematron is the branch, you might take a look if you get a chance).

We might not be able to implement it at this stage, but its a good target to work towards anyway I reckon.

**Comment by ballsteve** [comment]
Hi Daniel,

Sounds fantastic! I’ve used SchemaTron extensively when working for a previous client, so I’m a big fan.

Regarding the error, that looks like a RefMut problem. Let me know how you go with it, and I’ll also take a look if it can’t be resolved.

I’m currently working on adding some features to the XSLT module, such as current() and key() functions. Let me know if anything else needs to be added for Schematron.

Cheers,
Steve

**Comment by Devasta** [comment]
To be honest, I've never used Schematron before now, but not that I've had any aversion to it, just that I could right click in IntelliJ to generate any XSD I needed. 😅 

The error I am getting is probably user error, will let you know how I get on.
